File: backend/app/services/mock_predictor.py
import hashlib
import random
import time
from typing import Any


def _extract_features(rows: list[dict[str, Any]]) -> tuple[float, float, float]:
    capacity = [float(r["capacity_ah"]) for r in rows]
    early = sum(capacity[:50]) / min(50, len(capacity))
    rate = (capacity[0] - capacity[-1]) / len(capacity)
    return early, rate, capacity[0]


def _normalize_breakdown(e: int, m: int, c: int, q: int) -> dict[str, int]:
    total = e + m + c + q
    if total == 0:
        return {"electrical": 25, "mechanical": 30, "chemical": 35, "coulombic": 10}
    scaled = {
        "electrical": round(e * 100 / total),
        "mechanical": round(m * 100 / total),
        "chemical": round(c * 100 / total),
        "coulombic": round(q * 100 / total),
    }
    scaled["chemical"] += 100 - sum(scaled.values())
    return scaled


def _interpolate(anchors: list[tuple[float, float]], target: float) -> float:
    if target <= anchors[0][0]:
        return anchors[0][1]
    if target >= anchors[-1][0]:
        return anchors[-1][1]
    for (c0, v0), (c1, v1) in zip(anchors, anchors[1:]):
        if c0 <= target <= c1:
            span = c1 - c0
            return v0 if span == 0 else v0 + ((target - c0) / span) * (v1 - v0)
    return anchors[-1][1]


def _build_curve(cycle_life: int, cap500: float, cap1000: float) -> list[dict[str, float]]:
    anchors: list[tuple[float, float]] = [(0, 100.0)]
    if cycle_life >= 500:
        anchors.append((500, cap500))
    if cycle_life >= 1000:
        anchors.append((1000, cap1000))
    if anchors[-1][0] != cycle_life:
        anchors.append((cycle_life, max(75.0, cap1000 - 5)))

    step = max(50, cycle_life // 25)
    curve: list[dict[str, float]] = []
    for cycle in range(0, cycle_life + 1, step):
        capacity = _interpolate(anchors, cycle)
        band = 1 + (cycle / max(cycle_life, 1)) * 2
        curve.append(
            {
                "cycle": cycle,
                "capacity": round(capacity, 2),
                "lower": round(max(75, capacity - band), 2),
                "upper": round(min(100, capacity + band), 2),
            }
        )
    if curve[-1]["cycle"] != cycle_life:
        capacity = _interpolate(anchors, cycle_life)
        curve.append(
            {
                "cycle": cycle_life,
                "capacity": round(capacity, 2),
                "lower": round(max(75, capacity - 3), 2),
                "upper": round(min(100, capacity + 3), 2),
            }
        )
    return curve


def generate_prediction(rows: list[dict[str, Any]], battery_label: str | None = None) -> dict[str, Any]:
    time.sleep(1.2)

    early, rate, initial = _extract_features(rows)
    cycle_life = round(800 + early * 200 - rate * 5000)
    cycle_life = max(300, min(3000, cycle_life))

    norm = 100 / initial if initial > 0 else 1
    capacity_at_500 = round(early * 0.92 * norm, 1)
    capacity_at_1000 = round(early * 0.85 * norm, 1)
    ci_low = round(early * 0.82 * norm, 1)
    ci_high = round(early * 0.88 * norm, 1)

    rng = random.Random(hash(str(rows[:5]) + (battery_label or "")))
    degradation_breakdown = _normalize_breakdown(
        20 + rng.randint(0, 10),
        25 + rng.randint(0, 10),
        30 + rng.randint(0, 10),
        5 + rng.randint(0, 10),
    )

    seed = str(rows[:5]) + (battery_label or "")
    prediction_id = "pred_" + hashlib.sha256(seed.encode()).hexdigest()[:8]

    return {
        "prediction_id": prediction_id,
        "cycle_life": cycle_life,
        "capacity_at_500": capacity_at_500,
        "capacity_at_1000": capacity_at_1000,
        "confidence_interval": (ci_low, ci_high),
        "degradation_breakdown": degradation_breakdown,
        "capacity_curve": _build_curve(cycle_life, capacity_at_500, capacity_at_1000),
    }

File: frontend/src/lib/mockPredictor.ts
import type { CapacityPoint, CsvRow, PredictionResponse } from '@/types/prediction'

function extractFeatures(rows: CsvRow[]) {
  const capacity = rows.map((r) => r.capacity_ah)
  const early = capacity.slice(0, 50).reduce((a, b) => a + b, 0) / Math.min(50, capacity.length)
  const rate = (capacity[0] - capacity[capacity.length - 1]) / capacity.length
  return { early, rate, initial: capacity[0] }
}

function normalizeBreakdown(e: number, m: number, c: number, q: number) {
  const total = e + m + c + q
  if (total === 0) return { electrical: 25, mechanical: 30, chemical: 35, coulombic: 10 }
  const scaled = {
    electrical: Math.round((e * 100) / total),
    mechanical: Math.round((m * 100) / total),
    chemical: Math.round((c * 100) / total),
    coulombic: Math.round((q * 100) / total),
  }
  scaled.chemical += 100 - (scaled.electrical + scaled.mechanical + scaled.chemical + scaled.coulombic)
  return scaled
}

function interpolate(anchors: [number, number][], target: number) {
  if (target <= anchors[0][0]) return anchors[0][1]
  if (target >= anchors[anchors.length - 1][0]) return anchors[anchors.length - 1][1]
  for (let i = 0; i < anchors.length - 1; i++) {
    const [c0, v0] = anchors[i]
    const [c1, v1] = anchors[i + 1]
    if (c0 <= target && target <= c1) {
      const span = c1 - c0
      return span === 0 ? v0 : v0 + ((target - c0) / span) * (v1 - v0)
    }
  }
  return anchors[anchors.length - 1][1]
}

function buildCurve(cycleLife: number, cap500: number, cap1000: number): CapacityPoint[] {
  const anchors: [number, number][] = [[0, 100]]
  if (cycleLife >= 500) anchors.push([500, cap500])
  if (cycleLife >= 1000) anchors.push([1000, cap1000])
  if (anchors[anchors.length - 1][0] !== cycleLife) {
    anchors.push([cycleLife, Math.max(75, cap1000 - 5)])
  }

  const step = Math.max(50, Math.floor(cycleLife / 25))
  const curve: CapacityPoint[] = []
  for (let cycle = 0; cycle <= cycleLife; cycle += step) {
    const capacity = interpolate(anchors, cycle)
    const band = 1 + (cycle / Math.max(cycleLife, 1)) * 2
    curve.push({
      cycle,
      capacity: Math.round(capacity * 100) / 100,
      lower: Math.round(Math.max(75, capacity - band) * 100) / 100,
      upper: Math.round(Math.min(100, capacity + band) * 100) / 100,
    })
  }
  if (curve[curve.length - 1].cycle !== cycleLife) {
    const capacity = interpolate(anchors, cycleLife)
    curve.push({
      cycle: cycleLife,
      capacity: Math.round(capacity * 100) / 100,
      lower: Math.round(Math.max(75, capacity - 3) * 100) / 100,
      upper: Math.round(Math.min(100, capacity + 3) * 100) / 100,
    })
  }
  return curve
}

async function hashPredictionId(seed: string): Promise<string> {
  const buf = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(seed))
  const hex = Array.from(new Uint8Array(buf))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('')
  return `pred_${hex.slice(0, 8)}`
}

/** Client-side mock when the FastAPI service is unreachable (same logic as backend). */
export async function generateMockPrediction(
  rows: CsvRow[],
  batteryLabel?: string,
): Promise<PredictionResponse> {
  await new Promise((r) => setTimeout(r, 1200))

  const { early, rate, initial } = extractFeatures(rows)
  let cycleLife = Math.round(800 + early * 200 - rate * 5000)
  cycleLife = Math.max(300, Math.min(3000, cycleLife))

  const norm = initial > 0 ? 100 / initial : 1
  const capacity_at_500 = Math.round(early * 0.92 * norm * 10) / 10
  const capacity_at_1000 = Math.round(early * 0.85 * norm * 10) / 10
  const ci_low = Math.round(early * 0.82 * norm * 10) / 10
  const ci_high = Math.round(early * 0.88 * norm * 10) / 10

  const degradation_breakdown = normalizeBreakdown(
    20 + Math.floor(Math.random() * 11),
    25 + Math.floor(Math.random() * 11),
    30 + Math.floor(Math.random() * 11),
    5 + Math.floor(Math.random() * 11),
  )

  const seed = JSON.stringify(rows.slice(0, 5)) + (batteryLabel ?? '')
  const prediction_id = await hashPredictionId(seed)

  return {
    prediction_id,
    cycle_life: cycleLife,
    capacity_at_500,
    capacity_at_1000,
    confidence_interval: [ci_low, ci_high],
    degradation_breakdown,
    capacity_curve: buildCurve(cycleLife, capacity_at_500, capacity_at_1000),
  }
}
